### PR TITLE
[Task] added pages of friends

### PR DESCRIPTION
**Pull request by skrebetsnataliia**
https://github.com/ita-social-projects/GreenCity/issues/1957;
https://github.com/ita-social-projects/GreenCity/issues/995

added button for finding friends (for user without friends)
image

added displaying SixFriends and link to all-friends page

added friend's tab
image

page all-friends

page search (recommended friends)
image

request-item

if user don't has requests:

### CODE REVIEW

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+@Component({
+  selector: 'app-all-friends',
+  templateUrl: './all-friends.component.html',
+  styleUrls: ['./all-friends.component.scss']
+})
+export class AllFriendsComponent implements OnInit, OnDestroy {
+
+  public userId: number;
+  public Friends: FriendModel[];
+  private destroy = new Subject();
```
Comment (luchward) [inline src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts:18]: As I understand, this construction should be:
private destroy$: ReplaySubject<boolean> = new ReplaySubject(1);

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  public deleteFriend(id: number) {
+    this.userFriendsService.deleteFriend(this.userId, id).pipe(
+      catchError((error) => {
+        this.snackBar.openSnackBar('error.message');
+        return error;
+      })
+    )
+    .pipe(
+      takeUntil(this.destroy)
+    )
```
Comment (luchward) [inline src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts:90]: Some operators in a pipe are separated by comma: 
.pipe(
  operator1,
  operator2,
  ...
)

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  public deleteFriend(id: number) {
+    this.userFriendsService.deleteFriend(this.userId, id).pipe(
+      catchError((error) => {
+        this.snackBar.openSnackBar('error.message');
+        return error;
+      })
+    )
+    .pipe(
+      takeUntil(this.destroy)
```
Comment (luchward) [inline src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts:89]: I am not sure, but if you are interested in only one result, you should use take(1) operator.

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    });
+  }
+
+  public changeStatus(id) {
+    const index = this.Friends.findIndex(elem => elem.id === id);
+    this.Friends[index].added = !this.Friends[index].added;
+  }
+
+  public getSixFriends() {
+    this.userFriendsService.getAllFriends(this.userId).pipe(
```
Comment (luchward) [inline src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts:44]: In my view catching errors is implemented by interseptor. It invokes snackbar for displaying error messages.

**Review by luchward — APPROVED** [review]

**Review by mr-pavlenko — CHANGES_REQUESTED** [review]
Add please more description with screenshots.

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts (hunk 4 above)
Comment (skrebetsnataliia) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts:45]: Interseptor doen't catch all errors.

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts (hunk 3 above)
Comment (skrebetsnataliia) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts]: I think, that takeUntil is better in this case.

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts (hunk 2 above)
Comment (skrebetsnataliia) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts:91]: I've fixed it.

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts (hunk 1 above)
Comment (skrebetsnataliia) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts]: On my view construction new Subject() is posible too.

**Review by JimHolman1199 — APPROVED** [review]

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts (hunk 4 above)
Comment (luchward) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts:45]: In this case add new entry in error type object in the snackbar component

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.html [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<div class="friends-list">
+  <div class="container">
+    <div *ngFor="let friend of Friends">
```
Comment (mr-pavlenko) [inline src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.html:3]: Why `Friends` use upper case?

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    )
+    .subscribe (
+      (data: FriendArrayModel) => {
+        this.Friends = data.page;
+      },
+    );
+  }
+
+  public deleteFriendsFromList(id, array) {
+    const i = array.findIndex(item => item.id === id);
```
Comment (mr-pavlenko) [inline src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts:42]: what does variable `i` mean?

File: src/app/component/user/components/profile/users-friends/friend-dashboard/friend-dashboard.component.scss [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+      font-weight: 500;
+      font-size: 48px;
+      line-height: 48px;
+      font-family: 'Montserrat', sans-serif;
+    }
+  }
+
+  a {
+    color: $linkColorGray;
+    margin: 0 2px;
```
Comment (mr-pavlenko) [inline src/app/component/user/components/profile/users-friends/friend-dashboard/friend-dashboard.component.scss:32]: Why do you need 2px here?

File: src/app/component/user/components/profile/users-friends/friend-dashboard/friend-dashboard.component.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    this.translate.setDefaultLang(lang);
+  }
+
+  private subscribeToLangChange(): void {
+    this.langChangeSub = this.localStorageService.languageSubject
+      .subscribe(this.bindLang.bind(this));
+  }
+
+  ngOnDestroy(): void {
+    this.langChangeSub.unsubscribe();
```
Comment (mr-pavlenko) [inline src/app/component/user/components/profile/users-friends/friend-dashboard/friend-dashboard.component.ts:39]: You have 2 subscriptions and only one unsubscribe method.

File: src/app/component/user/components/profile/users-friends/friend-dashboard/friend-item/friend-item.component.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+})
+export class FriendItemComponent implements OnInit {
+
+  @Input() friend: FriendModel;
+  @Input() btnName: string;
+  @Output() friendEventEmit = new EventEmitter<number>();
+
+  constructor() { }
+
+  ngOnInit() {
```
Comment (mr-pavlenko) [inline src/app/component/user/components/profile/users-friends/friend-dashboard/friend-item/friend-item.component.ts:17]: Do you really need onInit there?

**Comment by mr-pavlenko** [comment]
Check the similar errors as described in the comments.

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.html (hunk 5 above)
Comment (yuriiLisunov) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.html]: fixed

File: src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts (hunk 6 above)
Comment (yuriiLisunov) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/all-friends/all-friends.component.ts]: fixed

File: src/app/component/user/components/profile/users-friends/friend-dashboard/friend-dashboard.component.scss (hunk 7 above)
Comment (yuriiLisunov) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/friend-dashboard.component.scss]: fixed

File: src/app/component/user/components/profile/users-friends/friend-dashboard/friend-dashboard.component.ts (hunk 8 above)
Comment (yuriiLisunov) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/friend-dashboard.component.ts:43]: fixed

File: src/app/component/user/components/profile/users-friends/friend-dashboard/friend-item/friend-item.component.ts (hunk 9 above)
Comment (yuriiLisunov) [reply src/app/component/user/components/profile/users-friends/friend-dashboard/friend-item/friend-item.component.ts]: fixed

**Review by idzhersh — APPROVED** [review]

**Review by yuriiLisunov — APPROVED** [review]

**Review by VitalyBorysenko — CHANGES_REQUESTED** [review]
bad displaing in 320 resolution

image

**Comment by yuriiLisunov** [comment]
I know about it. There will be a lot of changes (fix styles) when we will have all friends mockups for 320 and other resolutions

**Review by VitalyBorysenko — APPROVED** [review]

**Review by JimHolman1199 — APPROVED** [review]

**Review by ZooMingGG — APPROVED** [review]

**Review by mr-pavlenko — APPROVED** [review]

**Pull request merged**